dqn.py

```diff
@@ -150,48 +154,56 @@ def save_checkpoint(self, name):
         torch.save(self.Q_model, name)
 
 
-if __name__ == "__main__":
-    # and it's still not reproducible...
-    random_seed = 42
-    random.seed(random_seed)
-    np.random.seed(random_seed)
-    torch.manual_seed(random_seed)
+@dataclass
+class RLConfig:
+    seed: int = field(default=42, metadata="random seed for the environments")
+    env: str = field(default="CartPole-v1", metadata="the environment to run experiments on")
+    batch_size: int = field(default=32, metadata="the batch size of to train DQN")
+    sample_memory: bool =field(default=True, metadata="whether to use memory replay or correlated samples")
+    hidden_sizes: List[int] = field(default_factory=lambda: [128], metadata="list of hidden layer dimensions for DQN")
+    num_episodes: int = field(default=200, metadata="number of episodes to run DQN for")
+    epsilon: float = field(default=1.0, metadata="the change ot picking a random action")
+    discount_factor: float =field(default=0.8, metadata="discount over future rewards")
+    lr: float = field(default=1e-3, metadata="learning rate to train DQN")
+    use_target_net: bool = field(default=True, metadata="whether to use target network")
+    update_target_freq: int = field(default=10000, metadata="frequency of updating the target net")
+
 
-    parser = argparse.ArgumentParser()
-    parser.add_argument('--env', default="CartPole-v1", type=str, help='the environment to run experiments on')
-    parser.add_argument('--batch_size', default=32, type=int, help='the batch size of DQN')
-    parser.add_argument('--sample_memory', default=True, type=bool, help='whether to use memory replay or correlated samples')
-    parser.add_argument('--hidden_sizes', default=[128], type=List[int], help='size of DQN hidden layers')
-    parser.add_argument('--num_episodes', default=200, type=int, help='number of episodes to run DQN')
-    parser.add_argument('--epsilon', default=1.0, type=float, help='the change ot picking a random action')
-    parser.add_argument('--discount_factor', default=0.8, type=float, help='how much to discount future rewards')
-    parser.add_argument('--lr', default=1e-3, type=float, help='learning rate of DQN')
-    parser.add_argument('--use_target_net', default=True, type=bool, help='whether to use target network')
-    parser.add_argument('--update_target_freq', default=10000, type=int, help='how frequently to update target net')
+# Registering RLConfig class to enable duck typing
+cs = ConfigStore.instance()
+cs.store(name="config", node=RLConfig)
+log = logging.getLogger(__name__)
 
-    args = parser.parse_args()
+@hydra.main(config_name="config")
+def main(config: RLConfig) -> None:
+    """ Runs a training experiment based on the given hydra configuration """
 
-    env = gym.envs.make(args.env)
+    print(f"Launched! Experiment logs available at {os.getcwd()}.")
```

Comment:
Is it really at the cwd? Does hydra change the cwd?

Reply:
yeah, each run gets its own "experiment" directory. This way we can do multiple runs in parallel (in principle) for hyperparameter sweeps.

Reply:
Yeah, that's what I'd expect. I was just surprised that it changes the cwd.